File: src/generate_activity_log.js
const fs = require('fs')
const moment = require('moment-timezone')
const data = require('../tmp/data.json')
const dataPrev = require('../tmp/data_prev.json')

const updateLogFile = './tmp/updatelog/log.json'
var updateLog = require('.' + updateLogFile)

var statewiseNew = data.statewise.reduce((arr, row) => {
  arr[row.state] = row
  return arr
}, {})

var confirmedText
var recoveredText
var deathText
var fullText = ''
var tgFullText = ''
var isChanged = false

var relStates = {}

function initRelStates (state) {
  if (!relStates[state]) {
    relStates[state] = {}
  }
}

dataPrev.statewise.forEach(element => {
  if (element.state === 'Total') {
    return
  }
  isChanged = false
  confirmedText = null
  recoveredText = null
  deathText = null
  var text = null
  if (parseInt(element.confirmed) < parseInt(statewiseNew[element.state].confirmed)) {
    var confirmedDiff = statewiseNew[element.state].confirmed - element.confirmed
    confirmedText = confirmedDiff + ' new case' + (confirmedDiff === 1 ? '' : 's')
    isChanged = true
  }
  if (parseInt(element.recovered) < parseInt(statewiseNew[element.state].recovered)) {
    var recoveredDiff = statewiseNew[element.state].recovered - element.recovered
    recoveredText = recoveredDiff + ' recover' + (recoveredDiff === 1 ? 'y' : 'ies')
    isChanged = true
  }
  if (parseInt(element.deaths) < parseInt(statewiseNew[element.state].deaths)) {
    var deathDiff = statewiseNew[element.state].deaths - element.deaths
    deathText = deathDiff + ' death' + (deathDiff === 1 ? '' : 's')
    isChanged = true
  }

  if (isChanged) {
    text = (confirmedText ? confirmedText + ', ' : '') + (recoveredText ? recoveredText + ', ' : '') + (deathText ? deathText + ', ' : '')
    var arr = text.split(', ')
    if (arr.length > 2) {
      arr = text.split(', ')
      arr = arr.slice(0, -1)
      var arrLast = arr[arr.length - 1]
      arr = arr.slice(0, -1)
      text = arr.join(', ')
      text = text + ' and ' + arrLast
    } else {
      arr = arr.slice(0, -1)
      text = arr.join()
    }
    text = text + ' in ' + element.state
    fullText = fullText + text + '\n'
  }
})
function fillSpace (str, width) {
  var empt = Array(width - str.length).join(' ')
  return empt + str
}

const widthState = 2
const widthConfirmed = 15
const widthRecovered = 15
const widthDeceased = 14

function editMessage (lastUpdated) {
  data.statewise.forEach(element => {
    var stateCode = element.statecode
    if (stateCode === 'TT' || statewiseNew[element.state].confirmed === 0) {
      return
    }
    initRelStates(stateCode)
    relStates[stateCode].C = +statewiseNew[element.state].confirmed
    relStates[stateCode].Cd = +statewiseNew[element.state].deltaconfirmed
    relStates[stateCode].R = +statewiseNew[element.state].recovered
    relStates[stateCode].Rd = +statewiseNew[element.state].deltarecovered
    relStates[stateCode].D = +statewiseNew[element.state].deaths
    relStates[stateCode].Dd = +statewiseNew[element.state].deltadeaths
  })
  var words = fillSpace('St', widthState) +
        fillSpace('Cnfrmd', widthConfirmed) +
        fillSpace('Rcvrd', widthRecovered) +
        fillSpace('Dcsd', widthDeceased) + '\n'

  const lengthOfLine = widthState + widthConfirmed + widthRecovered + widthDeceased
  words += Array(lengthOfLine).join('-') + '\n'
  for (var element in relStates) {
    var c = '(' + relStates[element].Cd + ') ' + relStates[element].C
    var r = '(' + relStates[element].Rd + ') ' + relStates[element].R
    var d = '(' + relStates[element].Dd + ') ' + relStates[element].D

    words += fillSpace(element, widthState) +
            fillSpace(c, widthConfirmed) +
            fillSpace(r, widthRecovered) +
            fillSpace(d, widthDeceased) + '\n'
    // console.log(rel_states[element]);
  }
  var indiaTotal = '*Covid-19 India*\n'
  indiaTotal += 'Last updated: _' + lastUpdated + '_\n\n'

  var total = statewiseNew.Total
  indiaTotal += '```    Total cases: (↑' + total.deltaconfirmed + ') ' + total.confirmed +
        '\n' + '    Recovered  : (↑' + total.deltarecovered + ') ' + total.recovered +
        '\n' + '    Deaths     : (↑' + total.deltadeaths + ') ' + total.deaths + '```'

  // console.log(india_total);

  words = indiaTotal + '\n\n```\n' + words + '```\n\n*www.covid19india.org*'

  console.log(words)
  fs.writeFileSync('/tmp/apidata_iutable', words)
}

if (fullText !== '') {
  var total = statewiseNew.Total
  tgFullText = fullText + '\n' +
        '``` Total cases: (↑' + total.deltaconfirmed + ') ' + total.confirmed +
        '\n' + ' Recovered  : (↑' + total.deltarecovered + ') ' + total.recovered +
        '\n' + ' Deaths     : (↑' + total.deltadeaths + ') ' + total.deaths + '```'

  const now = moment().unix()
  var entry = {}
  entry.update = fullText
  entry.timestamp = now
  updateLog.push(entry)
  updateLog = updateLog.slice(-50)

  fs.writeFileSync(updateLogFile, JSON.stringify(updateLog, null, 2))

  var date = moment.unix(now)
  var formatedTime = date.tz('Asia/Kolkata').format('MMMM DD, hh:mm A') + ' IST'
  editMessage(formatedTime)

  // console.log(formated_time)
  var finalText = '_' +
        formatedTime + '_\n\n' +
        tgFullText +
        '\n\n*www.covid19india.org*'
  console.log(finalText)

  fs.writeFileSync('/tmp/apidata_iumessage', finalText)
} else {
  console.log('No updates this time!')
}

File: src/generate_activity_log.py
#!/usr/bin/env python3

import csv
import logging
import sys
from datetime import datetime
from pathlib import Path
import json
import time

ROOT_DIR = Path("tmp")
CSV_DIR = ROOT_DIR / "csv" / "latest"
STATE_WISE = ROOT_DIR / "csv" / "latest" / "state_wise.csv"
STATE_WISE_PREV = ROOT_DIR / "state_wise_prev"

# Set logging level
logging.basicConfig(stream=sys.stdout, format="%(message)s", level=logging.INFO)

data = dict()

total = dict()


if __name__ == "__main__":
    logging.info("generate_activity_log.py start")
    with open(STATE_WISE_PREV) as f:
        reader = csv.DictReader(f)
        for row in reader:
            if row["State"] not in data:
                if row["State"] == "Total":
                    pass
                else:
                    data[row["State"]] = {
                        "prev": {
                            "Confirmed": int(row["Confirmed"]),
                            "Recovered": int(row["Recovered"]),
                            "Deaths": int(row["Deaths"]),
                        },
                        "new": {
                            "Confirmed": 0,
                            "Recovered": 0,
                            "Deaths": 0,
                        },
                        "code": row["State_code"]
                    }
    with open(STATE_WISE) as f:
        reader = csv.DictReader(f)
        for row in reader:
            if row["State"] == "Total":
                total["Confirmed"] = row["Confirmed"]
                total["Recovered"] = row["Recovered"]
                total["Deaths"] = row["Deaths"]
                total["DeltaConfirmed"] = row["Delta_Confirmed"]
                total["DeltaRecovered"] = row["Delta_Recovered"]
                total["DeltaDeaths"] = row["Delta_Deaths"]
                continue
            if row["State"] not in data:
                logging.error("New state?")
                logging.error(row["State"])

            data[row["State"]]["new"]["Confirmed"] = int(row["Confirmed"])
            data[row["State"]]["new"]["Recovered"] = int(row["Recovered"])
            data[row["State"]]["new"]["Deaths"] = int(row["Deaths"])

    logging.info("Done------")

    longtext = ""
    updatelogtxt = ""
    # calc diff
    for state in data:
        tmptext = ""
        # logging.info(state)
        deltaConfirmed = (
            data[state]["new"]["Confirmed"] - data[state]["prev"]["Confirmed"]
        )
        deltaRecovered = (
            data[state]["new"]["Recovered"] - data[state]["prev"]["Recovered"]
        )
        deltaDeaths = data[state]["new"]["Deaths"] - data[state]["prev"]["Deaths"]

        if deltaConfirmed or deltaRecovered or deltaDeaths:
            # logging.info(state)
            if deltaConfirmed > 0:
                tmptext += str(deltaConfirmed) + " new confirmed cases\n"
            if deltaRecovered > 0:
                tmptext += str(deltaRecovered) + " new recoveries\n"
            if deltaDeaths > 0:
                tmptext += str(deltaDeaths) + " new deaths\n"
            longtext += "(#"+data[state]["code"]+") *" + state + ":*\n" + tmptext + "\n"
            updatelogtxt += "" + state + ":\n" + tmptext + "\n"
    if updatelogtxt:
        # logging.info(json.dumps(data, indent=3))
        current_time = datetime.now().strftime("%Y %b %d, %I:%M %p IST")
        longtext = "_" + current_time + "_\n\n" + longtext
        longtext += f"""```
 Total Cases: (↑{total["DeltaConfirmed"]}) {total["Confirmed"]}
 Recovered  : (↑{total["DeltaRecovered"]}) {total["Recovered"]}
 Deaths     : (↑{total["DeltaDeaths"]}) {total["Deaths"]}```

www.covid19india.org"""
        logging.info("generate updatelog/log.json")
        logging.info(longtext)
        with open("/tmp/apidata_iumessage", "a") as the_file:
            the_file.write(longtext)

        # logging.info(updatelogtxt)
        with open("./tmp/updatelog/log.json", "r") as f:
            data = json.load(f)
        with open("./tmp/updatelog/log.json", "w") as f:
            data.pop(0)
            timestamp = time.time()
            new_entry = {"update": updatelogtxt, "timestamp": int(timestamp)}
            data.append(new_entry)
            f.write(json.dumps(data, indent=3))

    else:
        logging.info("No delta changes")
    logging.info("generate_activity_log.py end")
